src/Microsoft.Data.Analysis.Interactive/DataFrameKernelExtension.cs

```diff
@@ -1,47 +1,140 @@
+            }
+
+            stringBuilder.AppendLine();
+            stringBuilder.Append(@"}");
+            prettyFormatter.AppendLine();
+            prettyFormatter.Append(@"}");
+            prettyFormatter.AppendLine();
+            return stringBuilder.ToString();
+        }
+
+        public Task OnLoadAsync(IKernel kernel)
+        {
             var kernelBase = kernel as KernelBase;
-            var directive = new Command("#!doit")
+            var directive = new Command("#!AddStrongTypes")
             {
-                Handler = CommandHandler.Create(async (FileInfo csv, string typeName, KernelInvocationContext context) =>
+                Handler = CommandHandler.Create(async (string dataFrameName, string typeName, KernelInvocationContext context) =>
                 {
                     // do the job
-                    var command = new SubmitCode(@$"public class {typeName}{{}}");
-                    context.Publish(new DisplayedValueProduced($"emitting {typeName} from {csv.FullName}", context.Command));
-                    await context.HandlingKernel.SendAsync(command);
+                    try
+                    {
+                        if (context.HandlingKernel is CSharpKernel cSharp)
+                        {
+                            System.Collections.Immutable.ImmutableArray<CodeAnalysis.Scripting.ScriptVariable> variables = cSharp.ScriptState.Variables;
+                            for (int i = 0; i < variables.Length; i++)
+                            {
+                                CodeAnalysis.Scripting.ScriptVariable variable = variables[i];
+                                if ((dataFrameName == null || variable.Name == dataFrameName) && variable.Value is DataFrame df)
+                                {
+                                    string typedDataFrame = GetTypedDataFrameWithProperties(df, typeName, out StringBuilder prettyFormatter);
+                                    context.Publish(new DisplayedValueProduced($"Created type {typeName}: ", context.Command));
+                                    await context.DisplayAsync(prettyFormatter.ToString());
+                                    var command = new SubmitCode(typedDataFrame);
+                                    await context.HandlingKernel.SendAsync(command);
+                                }
+                            }
+                        }
+                    }
+                    catch (Exception)
+                    {
+                        await context.DisplayAsync($"Encountered an exception. Could not create type {typeName}");
+                    }
+
                 })
             };
 
-            directive.AddOption(new Option<FileInfo>(
-                "csv").ExistingOnly());
-
             directive.AddOption(new Option<string>(
-                "typeName",
-                getDefaultValue:() => "Foo"));
+                "--type-name",
+                getDefaultValue: () => "InteractiveDataFrame"));
```

Comment:
Wouldn't using a constant as the default value name result in shadowing of any previous `InteractiveDataFrame` instances? The default should probably be qualified by some kind of cell identifier.

Reply:
It does result in shadowing but I figured that would be the expected behavior, since everything else in a notebook behaves this way. Adding a cell identifier would make that more obvious, you're right. I'll look into what identifier information is available here and update this code.

Reply:
This is tricky. Cell identifiers aren't something you can consistently get access to across Juptyer frontends, despite various workarounds that exist to accomplish this. Remember that cells are transient. Submissions are more important in the interactive session, and the types will align to the specific submissions. In some languages you won't be able to access shadowed types except via closure, so this can also be hard to reason about.